extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/photos/MicrosoftPhotosExporter.java

```diff
@@ -0,0 +1,253 @@
+  private static final String ODATA_NEXT = "@odata.nextLink";
+  private final String photosRootUrl;
+  private final String photosFolderTemplate;
+  private final String photosContentTemplate;
+
+  private final OkHttpClient client;
+  private final ObjectMapper objectMapper;
+  private final JobStore jobStore;
+
+  public MicrosoftPhotosExporter(
+      OkHttpClient client, ObjectMapper objectMapper, JobStore jobStore) {
+    this("https://graph.microsoft.com", client, objectMapper, jobStore);
+  }
+
+  @VisibleForTesting
+  public MicrosoftPhotosExporter(
+      String baseUrl, OkHttpClient client, ObjectMapper objectMapper, JobStore jobStore) {
+    photosRootUrl = baseUrl + "/v1.0/me/drive/special/photos/children";
+    photosFolderTemplate = baseUrl + "/v1.0/me/drive/items/%s/children";
+    photosContentTemplate = baseUrl + "/v1.0/me/drive/items/%s/content";
+    this.client = client;
+    this.objectMapper = objectMapper;
+    this.jobStore = jobStore;
+  }
+
+  @Override
+  public ExportResult<PhotosContainerResource> export(TokenAuthData authData) {
+    // TODO replace this with the real job id passed via a param
+    UUID jobId = UUID.randomUUID();
+
+    try {
+
+      List<Map<String, Object>> photoItems = requestAllPhotoItems(authData);
+
+      downloadAndCachePhotos(jobId, authData, photoItems);
+
+      PhotosContainerResource containerResource = convertToResource(photoItems);
+
+      return new ExportResult<>(ExportResult.ResultType.END, containerResource);
+
+    } catch (IOException e) {
+      e.printStackTrace(); // FIXME log error
+      return new ExportResult<>(
+          ExportResult.ResultType.ERROR, "Error retrieving contacts: " + e.getMessage());
+    }
+  }
+
+  @Override
+  public ExportResult<PhotosContainerResource> export(
+      TokenAuthData authData, ExportInformation exportInformation) {
+    return export(authData);
+  }
+
+  /**
+   * Recursively scans OneDrive folders starting with the special root photo folder. Returns all
+   * photo items, i.e. items with an image MIME type.
+   *
+   * <p>Folders are recursively processed using a deque of folder items. The deque is iterated, a
+   * dequeue folder item is popped, its contained items are requested, and any contained folders are
+   * added to the end of the dequeue. Processing completes when the dequeue has been exhausted.
+   *
+   * <p>All contained photo items are returned.
+   *
+   * @param authData the authorization data
+   */
+  private List<Map<String, Object>> requestAllPhotoItems(TokenAuthData authData)
+      throws IOException {
+    List<Map<String, Object>> photoItems = new ArrayList<>();
+
+    Deque<Map<String, Object>> folderItems = new ArrayDeque<>();
+    requestItems(photosRootUrl, authData, folderItems, photoItems);
+
+    // request items for all folders recursively until the collection is exhausted, adding contained
+    // folders to the end of the deque for processing
+    while (!folderItems.isEmpty()) {
```

Comment:
Rather than fetch everything in memory at once and store, the design we've evolved is to fetch the top level node with it's photos and any sub (nested) albums, then return that export result for import along with continuation data which then re-enters the copier, and so on recursively through the tree.   This is primarily for scalability but also for reliability when we add retry support.